filesystem/block.go

```diff
@@ -0,0 +1,340 @@
+	lastLength  uint64 // length of the last block
+}
+
+// newBlockRange returns the block range for 'size' bytes from 'from'.
+func newBlockRange(from, length uint64) blockRange {
+	end := from + length
+	return blockRange{
+		start:       int(from / BlockSize),
+		startOffset: from % BlockSize,
+		startLength: min(length, BlockSize-(from%BlockSize)),
+		last:        int(end / BlockSize),
+		lastLength:  end % BlockSize,
+	}
+}
+
+// shrink resizes the data to a smaller length.
+// Requirement: from > to.
+// If truncates are done on block boundaries, this is reasonably
+// efficient. However, if truncating in the middle of a block,
+// we need to fetch the block first, truncate it, and write it again.
+func shrink(e sqlExecutor, inodeID, from, to uint64) error {
+	delRange := newBlockRange(to, from-to)
+	deleteFrom := delRange.start
+
+	if delRange.startOffset > 0 {
+		// We're truncating in the middle of a block, fetch it, truncate its
+		// data, and write it again.
+		// TODO(marc): this would be more efficient if we had LEFT for bytes.
+		data, err := getBlockData(e, inodeID, delRange.start)
+		if err != nil {
+			return err
+		}
+		data = data[:delRange.startOffset]
+		if err := updateBlockData(e, inodeID, delRange.start, data); err != nil {
+			return err
+		}
+		// We don't need to delete this block.
+		deleteFrom++
+	}
+
+	deleteTo := delRange.last
+	if delRange.lastLength == 0 {
+		// The last block did not previously exist.
+		deleteTo--
+	}
+	if deleteTo < deleteFrom {
+		return nil
+	}
+
+	// There is something to delete.
+	// TODO(marc): would it be better to pass the block IDs?
+	delStmt := `DELETE FROM fs.BLOCK WHERE id = $1 AND block >= $2`
+	if _, err := e.Exec(delStmt, inodeID, deleteFrom); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// grow resizes the data to a larger length.
+// Requirement: to > from.
+// If the file ended in a partial block, we fetch it, grow it,
+// and write it back.
+func grow(e sqlExecutor, inodeID, from, to uint64) error {
+	addRange := newBlockRange(from, to-from)
+	insertFrom := addRange.start
+
+	if addRange.startOffset > 0 {
+		// We need to extend the original 'last block'.
+		// Fetch it, grow it, and update it.
+		// TODO(marc): this would be more efficient if we had RPAD for bytes.
+		data, err := getBlockData(e, inodeID, addRange.start)
+		if err != nil {
+			return err
+		}
+		data = append(data, make([]byte, addRange.startLength, addRange.startLength)...)
+		if err := updateBlockData(e, inodeID, addRange.start, data); err != nil {
+			return err
+		}
+		// We don't need to insert this block.
+		insertFrom++
+	}
+
+	insertTo := addRange.last
+	if insertTo < insertFrom {
+		return nil
+	}
+
+	// Build the sql statement and blocks to insert.
+	// We don't share this functionality with 'write' because we can repeat empty blocks.
+	// This would be shorter if we weren't trying to be efficient.
+	// TODO(marc): this would also be better if we supported sparse files.
```

Comment:
I think it would be simpler of you supported sparse files. A non-existent block for a file would indicate a block full of zeroes.